### PR TITLE
Refactor MSC implementation and add further tests

### PR DESCRIPTION
**Pull request by sethrj**
This changeset should be purely refactoring: the results should not change.
- Added several comments and clarifications to the MSC methods (e.g. `calc_eloss` -> `calc_stopping_energy` for the inverse range lookup)
- Removed `phys_step` from the  MSC interaction result
- Move a hardcoded parameter from the helper class to the main "MSC parameters" data class
- Precalculate more data in constructors for distributions (we may want to split up the actual sampling distribution as we did with energy loss fluctuations)
- Split out positron correction code into a helper class with additional testing
- Split geo/true and true/geo path length conversions into new classes and add lots of testing of different cases (including edge cases that require 'min' and 'clamp' and such)
- Move path length conversion out of step limiter and scatter sampling
- Remove local "geo step" in along-step that are no longer needed
- Allow array construction for polynomial evaluator

They have allowed me to identify a few issues that I have reverted and will restore in a separate "fix" MR:
- Clarified some of the conditionals in the MSC samplers which allow for removal of redundant checks
- Remove a superfluous Taylor expansion of $sin(\theta_0)$ (assume this was premature optimization on the developer's part)
- Replace another conditional expansion of $1 - \exp(x)$ with `expm1`
- Correct the symmetry of the geo/true step conversions so now they do exactly the same thing (including using `logp1` to balance `expm1`)
- Mark 1 MFP-limited steps as being MSC limited (currently their "action" might be incorrect)
- Use `alpha = 0` for small step approximation since that's literally the equivalent meaning (constant cross section)

With just the refactoring:
- `UrbanMsc::calc_step` (now `limit_step`) register usage drops from 40 to 26
- `UrbanMsc::apply_step` register usage drops from 45 to 38
- Surrogate along-step kernel resources do *not* change
- Total instruction count for surrogate along-step drops by 15%

See https://github.com/celeritas-project/cuda-test-snippets/commit/809e00b55b0f8dca4ad45b7df2e3997a1ad5e860  .

### CODE REVIEW

**Comment by sethrj** [comment]
Since @amandalund is working on #615 I can revert the `alpha` sentinel commit for her fix branch.

**Comment by whokion** [comment]
@sethrj Please fix the compilation error and a couple of warnings,  
`warning: calling a constexpr __host__ function("tiny_step_alpha") from a __host__ __device__ function("calc_true_path") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.`

File: src/celeritas/em/UrbanMscParams.cc [hunk 1]
```diff
@@ -200,7 +200,7 @@ UrbanMscParams::calc_material_data(MaterialView const& material_view)
 
     UrbanMscMaterialData data;
 
-    double zeff = material_view.zeff();
+    double const zeff = material_view.zeff();
```
Comment (stognini) [inline src/celeritas/em/UrbanMscParams.cc:203]: Is there a reason to have all of these as `double` instead of `real_type`?

Comment (sethrj) [reply src/celeritas/em/UrbanMscParams.cc:203]: Since it's all setup code on the CPU, and CPU floating point operations are no slower with double, we might as well compute with the extra precision before truncating? It probably doesn't matter in practice.

File: src/celeritas/em/msc/UrbanMscScatter.hh [hunk 2]
```diff
@@ -459,28 +514,36 @@ CELER_FUNCTION real_type UrbanMscScatter::simple_scattering(
...
     // Correction factor from e- scattering data
     theta0 *= (msc_.coeffth1 + msc_.coeffth2 * std::log(y));
 
+    if (true_path_ < limit_min_)
+    {
+        // Correct for non-MSC-limited path lengths
+        theta0 *= std::sqrt(true_path_ / limit_min_);
+    }
+
+    CELER_ENSURE(theta0 >= 0);
```
Comment (whokion) [inline src/celeritas/em/msc/UrbanMscScatter.hh:546]: This seems a hard condition as theta0 can be a very small negative value (for a very large radiation length, ~1e+17, galactic material?).

Comment (sethrj) [reply src/celeritas/em/msc/UrbanMscScatter.hh:546]: In that case, wouldn't the "skip_sampling" condition `true_path_ <= lambda_ * params_.tau_small` prevent us from even calculating theta0?

File: src/celeritas/em/msc/UrbanMscScatter.hh [hunk 3]
```diff
@@ -159,36 +157,115 @@ UrbanMscScatter::UrbanMscScatter(UrbanMscRef const& shared,
...
+            // TODO: this should be virtually impossible because of
+            // skip_sampling_ above
+            is_displaced_ = false;
+        }
+        else if (tau_ < params_.tau_big)
+        {
+            // TODO: if displacement is disabled (because of being near the
+            // boundary or having a short step) then limit_min_ will be zero,
+            // and some corrections will not be applied.
+            // CELER_ASSERT(limit_min_ > 0);
```
Comment (sethrj) [inline src/celeritas/em/msc/UrbanMscScatter.hh:257]: @whokion I noticed this unexpected behavior: in `UrbanMscStepLimit` we skip the calculation of `limit_min` (when entering a volume) for short steps. That means the initial step through a volume has `limit_min = 0` so `compute_theta0` doesn't account for the minimum limit. Fixing this doesn't seem to change any of the unit tests, though, so it's probably not important.

File: src/celeritas/em/msc/UrbanMscScatter.hh (hunk 2 above)
Comment (whokion) [reply src/celeritas/em/msc/UrbanMscScatter.hh:544]: It may be true (i.e, should be self-consistent), but the negative case is due to the material (radiation length) dependent correction to the apparently non-detector (galactic or gaseous?) volume of which radiation length is humongous (probably not physically meaningful for the msc scattering, but nothing practically wrong as it is close to 0-) while the skip_sampling condition is not always hold with the corresponding lambda (even though the m.p.f value is still big, but smaller than true_path * tau_small).  Anyway, as the theta0 itself is based on the 2nd momentum of the angular differential cross section, I do agree that it should be always positive, but should not impose the requirement due to the rather ah-hoc empirical correction (and the evaluated lambda for this non-detector material, whatever it is).

File: src/celeritas/em/msc/UrbanMscScatter.hh (hunk 3 above)
Comment (whokion) [reply src/celeritas/em/msc/UrbanMscScatter.hh:255]: Okay, then should we initialize limit_min_ with  something like`limit_min_fix() `as [this if block](https://github.com/sethrj/celeritas/blob/msc-units/src/celeritas/em/msc/UrbanMscStepLimit.hh#L143) may not completely cover all possible cases?

Comment (whokion) [reply src/celeritas/em/msc/UrbanMscScatter.hh:255]: As looking at [this line](https://geant4.kek.jp/lxr/source/processes/electromagnetic/standard/src/G4UrbanMscModel.cc#L97) or [this line](https://geant4.kek.jp/lxr/source/processes/electromagnetic/standard/src/G4UrbanMscModel.cc#L434) again, I confirm that the limit_min_ should be initialized with UrbanMscParameters::limit_min() which is apparently removed from this PR, but equivalently can be set to 10*limit_min_fix().

Comment (sethrj) [reply src/celeritas/em/msc/UrbanMscScatter.hh:255]: OK, I removed `limit_min` because it wasn't used anywhere: but I guess it should be 😄 so we now have the same behavior as Geant4.

File: src/celeritas/em/data/UrbanMscData.hh [hunk 4]
```diff
@@ -66,6 +60,12 @@ struct UrbanMscParameters
     {
         return units::MevEnergy{1e-6};
     }
+
+    //! The lower bound of energy to scale the minimum true path length limit
+    static CELER_CONSTEXPR_FUNCTION Energy min_scaling_energy()
```
Comment (whokion) [inline src/celeritas/em/data/UrbanMscData.hh:71]: The "scaling" is confusing as this energy is nothing to do with any scaling, but a cut off to evaluate the true path length limit.

File: src/celeritas/em/msc/UrbanMscHelper.hh [hunk 5]
```diff
@@ -50,17 +50,19 @@ class UrbanMscHelper
     // TODO: the following methods are used only by MscStepLimit
 
     // The total energy loss over a given step length
-    inline CELER_FUNCTION Energy calc_eloss(real_type step) const;
-
-    // The kinetic energy at the end of a given step length corrected by dedx
-    inline CELER_FUNCTION Energy calc_end_energy(real_type step) const;
+    inline CELER_FUNCTION Energy calc_stopping_energy(real_type step) const;
```
Comment (whokion) [inline src/celeritas/em/msc/UrbanMscHelper.hh:53]: Again, this name is confusing as it sounds like something related to energy loss itself rather than the energy at the end of step taking into account for the energy loss by the step length.

Comment (whokion) [reply src/celeritas/em/msc/UrbanMscHelper.hh:53]: Sorry. Never mind.

File: src/celeritas/em/msc/UrbanMscHelper.hh [hunk 6]
```diff
@@ -120,14 +122,16 @@ CELER_FUNCTION real_type UrbanMscHelper::calc_msc_mfp(Energy energy) const
 
 //---------------------------------------------------------------------------//
 /*!
- * Calculate the total energy loss over a given step length.
+ * Calculate the energy of a track that is stopped after the given step.
```
Comment (whokion) [inline src/celeritas/em/msc/UrbanMscHelper.hh:125]: I am lost here as this sounds like that the track is `stopped` at the end of step.

File: src/celeritas/em/msc/UrbanMscScatter.hh [hunk 7]
```diff
@@ -222,12 +324,14 @@ CELER_FUNCTION auto UrbanMscScatter::operator()(Engine& rng) -> MscInteraction
...
     {
-        // This should only be needed to silence compiler warning
+        // This should only be needed to silence compiler warning, since the
+        // displacement should be ignored since our action result is
+        // 'scattered'
         result.displacement = {0, 0, 0};
     }
 
     // Calculate displacement
-    if (is_displaced_ && tau_ >= params_.tau_small)
```
Comment (whokion) [inline src/celeritas/em/msc/UrbanMscScatter.hh:230]: So, is the second condition absorbed into is_displaced? Why?

File: src/celeritas/em/msc/UrbanMscStepLimit.hh [hunk 8]
```diff
@@ -81,11 +81,14 @@ class UrbanMscStepLimit
     // Urban MSC range properties
     MscRange msc_range_;
 
-    bool on_boundary_{};
```
Comment (whokion) [inline src/celeritas/em/msc/UrbanMscStepLimit.hh:84]: How is this condition (the first step of the tracking or the first step in the new volume) treated now?

File: src/celeritas/em/msc/UrbanMscStepLimit.hh [hunk 9]
```diff
@@ -164,53 +185,28 @@ UrbanMscStepLimit::UrbanMscStepLimit(UrbanMscRef const& shared,
...
-        }
-        msc_range_.limit_min = this->calc_limit_min();
-
-        // Store persistent range properties within this tracking volume
-        physics_.msc_range(msc_range_);
-    }
-    // The step limit
     real_type limit = range_;
-    if (limit > safety_)
+    if (range_ > safety_)
```
Comment (whokion) [inline src/celeritas/em/msc/UrbanMscStepLimit.hh:201]: Why is this change meaningfully better than limit > safety ?

**Review by whokion — COMMENTED** [review]
I thought that this PR is mainly for fixing debugging issues, not for major refactoring or optimizing codes (which seems too early as we still may directly compare code logic with Geant4 until the first physics validation is completed).  Also it is somewhat difficult to follow all motivations behinds detail changes, especially for UrbanMscScatter (which is of course the lack of thoughtful review from my side).

File: src/celeritas/em/UrbanMscParams.cc [hunk 10]
```diff
@@ -221,8 +221,12 @@ UrbanMscParams::calc_material_data(MaterialView const& material_view)
 
     // Parameters for the maximum distance that particles can travel
     data.d_over_r = 9.6280e-1 - 8.4848e-2 * std::sqrt(zeff) + 4.3769e-3 * zeff;
+    // XXX there must be another scaling factor missing here: how can the step
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/em/UrbanMscParams.cc:224]: Ah, it took me a minute to figure out what this comment was about. I might be confusing `d_over_r` with `r_over_range`: from the comments this looked like a heuristic maximum distance based on the particle's range? So I was confused how hadrons could have a longer distance than their range.

File: src/celeritas/em/data/UrbanMscData.hh [hunk 11]
```diff
@@ -49,12 +49,6 @@ struct UrbanMscParameters
         return 1e-9 * units::centimeter;
     }
 
-    //! Default minimum of the true path limit: 1e-8 cm
-    static CELER_CONSTEXPR_FUNCTION real_type limit_min()
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/em/data/UrbanMscData.hh:53]: This has been restored now that it's actually being used 😄

File: src/celeritas/em/data/UrbanMscData.hh (hunk 4 above)
Comment (sethrj) [reply src/celeritas/em/data/UrbanMscData.hh:71]: I was trying to convey that this is an energy cutoff for using an extra scaling factor in the `calc_limit_min` calculation.

File: src/celeritas/em/msc/UrbanMscHelper.hh (hunk 6 above)
Comment (sethrj) [reply src/celeritas/em/msc/UrbanMscHelper.hh:125]: Yeah I reworded this poorly: this is the energy of a hypothetical particle that, if it moves a distance `step`, has slowed down to zero energy at the end of the step.  Maybe we should just rename this to `calc_inverse_range`?

File: src/celeritas/em/msc/UrbanMscStepLimit.hh (hunk 8 above)
Comment (sethrj) [reply src/celeritas/em/msc/UrbanMscStepLimit.hh:84]: It's evaulated in the constructor rather than at `operator()` time, so we don't need the member data any more.

File: src/celeritas/em/msc/UrbanMscStepLimit.hh (hunk 9 above)
Comment (sethrj) [reply src/celeritas/em/msc/UrbanMscStepLimit.hh:201]: The `limit` is the result of the calculation; the conditional here isn't "is the limit greater than the safety" but "is the range greater than the safety". The two are of course identical at this point but I think the second conveys the intended meaning (and since the range doesn't change, it's arguably less error-prone).

File: src/celeritas/em/msc/UrbanMscScatter.hh (hunk 7 above)
Comment (sethrj) [reply src/celeritas/em/msc/UrbanMscScatter.hh:230]: This is equivalent to `tau_ < params_.tau_small` forcing `is_displaced_` to be `false`. This condition is now precalculated in the constructor because it changes the whole sampling strategy. (It's not affected by the random number selection so ideally shouldn't be inside the `operator()`.

**Comment by sethrj** [comment]
@amandalund Can you look at this at your leisure? I'll request @whokion's review on the (upcoming) proposed fixes.

File: src/celeritas/em/msc/UrbanMscScatter.hh (hunk 2 above)
Comment (sethrj) [reply src/celeritas/em/msc/UrbanMscScatter.hh:534]: Given this empirical model of theta0 that's implemented, the result should be strictly positive and it improves code clarity (and may catch a bug) if we mark it as such.

**Comment by sethrj** [comment]
OK I'm 100% done touching this now 😂

**Comment by amandalund** [comment]
I'll take a look today :)

**Comment by sethrj** [comment]
I am seeing some new errors in the regression suite:
```python
{'condition': 'theta0 > 0',
 'context': {'dir': [0.8533568023293286,
     0.5051686239543767,
     -0.1287898648586331],
     'energy': [0.30259283761261346, 'MeV'],
     'event': 6,
     'label': 'along-step-uniform-msc',
     'num_steps': 0,
     'parent': 10931837,
     'particle': 1,
     'pos': [-10.326360284713983,
     -0.6625442300126325,
     0.0862354958901772],
     'thread': 3910,
     'track': 10935693,
     'type': 'KernelContextException',
     'volume': 26},
 'file': 'src/celeritas/em/msc/detail/UrbanMscScatter.hh',
 'line': 537,
 'type': 'DebugError',
 'which': 'postcondition failed'}
```
and
```
src/corecel/math/Algorithms.hh:342: celeritas: internal assertion failed: a > 0 || (a == 0 && b != 0)
```
so I'll see what's up with that...

File: src/celeritas/em/msc/detail/MscStepToGeo.hh [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    {
+        // Eq 8.8: mfp_slope = (lambda_f / lambda_i) = (1 - alpha * tstep)
+        real_type mfp_slope;
+        if (energy_ < value_as<Mass>(shared_.electron_mass) || tstep == range_)
+        {
+            // Low energy or range-limited step
+            // (For range-limited step, the final cross section and thus MFP
+            // slope are zero).
+            result.alpha = 1 / range_;
+            mfp_slope = 1 - result.alpha * tstep;
```
Comment (amandalund) [inline src/celeritas/em/msc/detail/MscStepToGeo.hh:159]: I hit the assertion failure in `fastpow()` when the step was range-limited (`tstep == range_`) and the calculated `mfp_slope` was not exactly zero (e.g. -4.89081360221648e-17).

**Review by amandalund — COMMENTED** [review]
This is looking great @sethrj! On a first pass through I couldn't find any inconsistencies, but we should understand and fix the new assertion failures before merging. At least one of them looks like an issue with floating point accuracy.

File: src/celeritas/phys/Interaction.hh [hunk 13]
```diff
@@ -72,17 +72,22 @@ struct Interaction
  * - is_displaced == false ? limit_min is unchanged and alpha ==
  *   small_step_alpha()
  * - true_step >= geom_path
+ *
+ * The value \f$ \alpha \f$ is used in the approximation of the MSC
+ * transport cross section as a linear function over the current step. It is
+ * the negative slope of the transport MFP from start to stop, divided by the
+ * starting MFP. (Since MFP decreases with decreasing energy over the step,
+ * alpha should always be positive.)
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/phys/Interaction.hh:80]: Right, should definitely be an "if" and I'll note the cross section bug that violates the condition.

File: src/celeritas/em/msc/detail/MscStepFromGeo.hh [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+ *     t(z) = \langle t \rangle = -\lambda_{1} \log(1 - \frac{z}{\lambda_{1}})
+ * \f]
+ * or \f$ t(z) = \frac{1}{\alpha} [ 1 - (1-\alpha w z)^{1/w}] \f$ if the
+ * geom path is small, where \f$ w = 1 + \frac{1}{\alpha \lambda_{10}}\f$.
+ *
+ * \param true_path the proposed step before transportation.
+ * \param gstep the proposed step after transportation.
+ * \param alpha variable from UrbanMscStepLimit.
+ */
+class MscStepFromGeo
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/celeritas/em/msc/detail/MscStepFromGeo.hh:42]: From what I can tell, the two MSC models use slightly different implementations (variable names, coding styles) of the exact same algorithm. From the physics manual it seems like the path length conversion is a straightforward function of the transport mean free path length, although *that* value is dependent on the MSC models.

**Comment by sethrj** [comment]
OK, the `theta0>0` assertion fails because `limit_min_` wasn't getting set (a bugfix which I reverted to round out this PR), so the `true_path` (which should be the max of the physics path length and `limit_min_`) I think ends up being too small :
```
invalid theta0=-4.4752795550152e-06: true_path=6.495513259566e-09, limit_min_=0, y =9.0488720195809e-09, invbetacp=0.1715407860572
```

Hilariously: https://github.com/celeritas-project/celeritas/pull/641#discussion_r1124485671 .... @whokion 😉

File: src/celeritas/em/msc/detail/MscStepToGeo.hh (hunk 12 above)
Comment (sethrj) [reply src/celeritas/em/msc/detail/MscStepToGeo.hh]: Aha, thanks. I've fixed it with a `max`. Guess I forgot that with floating point arithmetic, `1 - (1/x) * x` isn't necessarily zero 😬

**Comment by sethrj** [comment]
Test failure is just a random MPI glitch! I don't see any more failures in MSC in reldeb mode, and the number of steps on those problems hasn't changed, suggesting the implementation changes here don't affect the answers.

File: src/celeritas/em/msc/detail/UrbanMscStepLimit.hh [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+
+    if (phys_step_ < shared_.params.limit_min_fix())
+    {
+        // Very short step: don't displace
+        skip_displacement_ = true;
+    }
+    else if (helper_.max_step() < safety_)
+    {
+        // The lower limit for the linear distance that the track can travel in
+        // one step is closer than the nearest boundary. Displacement should
```
Comment (amandalund) [inline src/celeritas/em/msc/detail/UrbanMscStepLimit.hh:140]: Upper limit?

**Review by amandalund — APPROVED** [review]
Yay, looks good to me @sethrj!
